Assignment 1 solution

emurray2 opened the pull request:
Here is my solution for assignment 1. I was a little bit confused on how to copy the blocks into the comb filter in the main file and save the output. I also still need to implement tests.

emurray2 committed:
Add RingBuffer solution code

emurray2 committed:
Add fields

emurray2 committed:
Add hash map for parameters

emurray2 committed:
Remove gain and delay as fields

Store in `parameters` hash map instead

emurray2 committed:
Initialize fields

emurray2 committed:
Set read and write locations of ring buffer

emurray2 committed:
Implement reset

emurray2 committed:
Add more fields

emurray2 committed:
Bug fixes

emurray2 committed:
Implement get_param

emurray2 committed:
Beginnings of the process function

emurray2 committed:
Finish implementing processing function

emurray2 committed:
Fix bug with capacity

emurray2 committed:
Read file by blocks

emurray2 committed:
Fixes buffers (kind of)

One of the channels is still not correct

emurray2 committed:
Add first test to Readme

TeAiris commented on src/comb_filter.rs:
```diff
@@ -21,23 +28,74 @@ pub enum Error {
+        for buffer in self.delay_line_list.iter_mut() {
+            buffer.reset();
+        }
     }
```
It could be helpful to add code to reset your delay and gain here. Or maybe separate functions for each that could be useful when linking to UI later. Ex. Double click knob to set to default value.

emurray2 committed:
Reset parameters to default value

emurray2 committed:
Merge branch 'assignment-1' of https://github.com/emurray2/ase-2024 into assignment-1

TeAiris commented on src/main.rs:
```diff
@@ -27,9 +30,44 @@ fn main() {
+        out.write_sample(0).unwrap();
+        i += 1;
     }
 }
```
I am a bit unsure but I wanted to share related to your statement about saving the output of the blocks.  
// If you use your code to process the blocks with comb_filter....
// Process input block with FIR comb filter
       
 comb_filter.process(input_buffer_slice_2d, output_buffer_slice_2d);

// You should be able to do something like this
// Write output block to the output text file
        
for (i, sample) in output_buffer.iter().enumerate() {
            write!(out, "{}{}", sample, if i % channels == channels - 1 { "\n" } else { " " }).unwrap();

The write out is similar to what we did in exercise 2. 

I used some of my code so you may need to revise based on your variable. I hope this is helpful.

emurray2 replied:
Yeah this one was a challenge for me. I was trying to figure out what to do with the remaining samples of the file since the file I used was 3 seconds long. 44100 * 3 is 132300 and not completely divisible by 1024. Thus, for now I set the remaining samples for each channel to zero. Ideally I would probably do another process block but with the remaining samples + padded zeros.

emurray2 replied:
I decided to check if there are remaining samples and then process another block with these remaining samples zero padded.

TeAiris reviewed (commented):
Everything is well written and easy to read! I will add more comments if I can once you add test functions.

emurray2 commented:
> Everything is well written and easy to read! I will add more comments if I can once you add test functions.

Thanks for the advice 🙏 ! Going to try and add some tests tonight when I get a chance. I'll keep you posted 🙌 !

mtsandra commented on src/main.rs:
```diff
@@ -27,9 +30,44 @@ fn main() {
+    let mut block_index: u32 = 0;
+    let mut input_buffer = vec![vec![f32::default(); channels as usize]; block_size as usize];
+    let mut output_buffer = vec![vec![f32::default(); channels as usize]; block_size as usize];
+    let mut comb_filter = CombFilter::new(FilterType::FIR, 1.0, 44100.0, channels as usize);
```
I think the assignment asked us to take the comb filter parameters from command line, so maybe it would be good to assign the variables using &args[i] instead of hard coding it here

mtsandra commented on src/main.rs:
```diff
@@ -27,9 +30,44 @@ fn main() {
+            let mut output_buffer_slice = output_buffer[..].iter_mut().map(|v| v.as_mut_slice()).collect::<Vec<_>>();
+            let output_buffer_slice_2d = &mut output_buffer_slice[..];
+            comb_filter.process(input_buffer_slice_2d, output_buffer_slice_2d);
+            for block_index in output_buffer_slice_2d.iter_mut() {
```
I'm wondering how you handled the edges where the delayed samples flow into the next block

emurray2 replied:
This could be the source of some error in the plot. Let me check and see if I'm understanding it correctly.

emurray2 replied:
I looked over the code and all the delayed samples are stored in ring buffers for each channel which are part of the `comb_filter.rs` file. That way they are separate from the input and output buffers defined here and still remembered at the next block.

mtsandra reviewed (commented):
I left specific questions and comments in line. Overall very clear, and I'll also add more comments once the tests are added!

emurray2 committed:
Add command line parameter arguments

emurray2 committed:
Handle remaining samples

Handle remaining samples to be processed by putting them in a zero-padded block

emurray2 committed:
Update plot and MATLAB code

emurray2 committed:
Update IIR Filter

emurray2 committed:
Add IIR Plot

emurray2 committed:
Recalculate delay length on set

emurray2 committed:
Add drumloop plots

emurray2 committed:
Create drumloop.wav

emurray2 committed:
Create sweep.wav

emurray2 committed:
Add test zero output

emurray2 committed:
Add test magnitude increase

emurray2 committed:
Add test varying block size

emurray2 committed:
Add test correct zero

emurray2 committed:
Add test mono and multichannel

emurray2 commented:
@TeAiris @mtsandra Alright, tests are done now. I know it's kind of last minute, but feel free to look over them and leave any more comments if you wish.

emurray2 committed:
Delete .idea/workspace.xml

emurray2 committed:
Add comment clarifying last test
